language/stdlib/modules/vector.mvir

```diff
@@ -22,8 +22,51 @@ module Vector {
   // Pop an element from the end of vector, abort if the vector is empty.
   native public pop_back<Element>(v: &mut Self.T<Element>): Element;
 
+  // Swaps the elements at the i'th and j'th indices in the vector.
+  native public swap<Element>(v: &mut Self.T<Element>, i: u64, j: u64);
+
+  // Reverses the order of the elements in the vector in place.
+  public reverse<Element>(v: &mut Self.T<Element>) {
+      let front_index: u64;
+      let back_index: u64;
+      let len: u64;
+      front_index = 0;
+      len = Self.length<Element>(freeze(copy(v)));
+
+      if (copy(len) == 0) {
+          return;
+      }
+      assert(copy(len) > 0, 0);
+
+      back_index = move(len) - 1;
+
+      while (copy(front_index) < copy(back_index)) {
+          Self.swap<Element>(copy(v), copy(front_index), copy(back_index));
+          front_index = move(front_index) + 1;
+          back_index = move(back_index) - 1;
+      }
+
+      return;
+  }
+
+  // Moves all of the elements of the `other` vector into the `lhs` vector.
+  // Leaves the `other` vector empty at the end of the call.
+  public append<Element>(lhs: &mut Self.T<Element>, other: &mut Self.T<Element>) {
```

Comment:
Should `other` be an owned vector instead? 

In current design, the empty vector that `other` points to after doesn't seem like it will be much use to the caller. In addition, it will need to be explicitly destroyed via a separate call whenever `Element` is a resource.

Reply:
We could make `other` owned, but I was following the convention that Rust has around this as well (where they take in a `&mut` instead). Happy to change it, but I think we should probably try to stay as close as possible to the rust API.

Reply:
I like `&mut T` more. The reason being that you can always get a `&mut T` from `T` but not the other way. The reason is that we don't have an equivalent of `std::mem::replace` in Move. Do you think we should provide that as a native? It seems reasonable to me as we can actually use it to implement swap.

Reply:
I suspect @sblackshear is implying that if we pass the vector by value (owned vector) the module will be able to destroy it at the end without having to tax the caller with any extra API.
So that is an interesting consideration.
Thoughts?

Reply:
I personally have a preference for passing in an owned vector. However, we should think carefully about how close (or not) we want to be to Rust on this.  Also, from an idiomatic-usage perspective being able to  re-use a vector after appending it is confusing.

For now, I propose: let's keep it mutable for now since that is more flexible, but if we want to later on we can change it.

Reply:
My thought is: when we're faced with a tradeoff between flexibility and predictability, we should choose predictability unless there's strong evidence that the flexibility is needed.

In this case:
(1) Being able to reuse the empty vector that gets returned doesn't buy you any flexibility (you could create yourself a new empty vector instead) and has the downside that you need to explicitly destroy it
(2) *But*, the flexibility to take a `&mut Vec.T` instead of a `Vec.T` might buy you something (e.g., maybe you want to suck all of the elements out of a vector published in a non-removable resource in global storage).

I'm not creative enough to think of a compelling use-case for (2), and (1) seems like a clear downside in terms of predictability and convenience. This makes me prefer the API that takes ownership of the second vector. But open to other arguments.

A note: I don't think consistency with Rust APIs is always a good goal for us to shoot for or a good argument for doing something in a certain way. The languages are different enough that we will often want to make different decisions. For example, the explicit destruction downside described in (1) doesn't exist in Rust because if is affine instead of linear.

Reply:
just a quick note as far as the consistency to Rust goes: immaterial! It's a non argument and a non starter in my opinion